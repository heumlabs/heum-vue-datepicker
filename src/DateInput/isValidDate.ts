import { Dayjs } from 'dayjs';

const isValidDate = (
  targetDate: Dayjs,
  disableDatesBefore?: Dayjs,
  disableDatesAfter?: Dayjs,
): boolean => {
  if (!targetDate.isValid()) {
    return false;
  } 

  if (
    disableDatesBefore 
    && disableDatesBefore.isValid() 
    && targetDate.isBefore(disableDatesBefore)
  ) {
    return false;
  }

  if (
    disableDatesAfter 
    && disableDatesAfter.isValid()
    && targetDate.isAfter(disableDatesAfter)
  ) {
    return false;
  }

  return true;
};

export default isValidDate;
