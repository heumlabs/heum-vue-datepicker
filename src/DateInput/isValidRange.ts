import { Dayjs } from 'dayjs';

const isInValidRange = (
  targetDate: Dayjs,
  disableDatesBefore?: Dayjs,
  disableDatesAfter?: Dayjs,
): boolean => {
  if (disableDatesBefore && targetDate.isBefore(disableDatesBefore)) {
    return false;
  }

  if (disableDatesAfter && targetDate.isAfter(disableDatesAfter)) {
    return false;
  }

  return true;
};

export default isInValidRange;
