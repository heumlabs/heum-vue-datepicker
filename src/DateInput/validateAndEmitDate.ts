import { DATE_FORMAT } from "@/constants";
import convertDate from "@/utils/convertDate";
import dayjs from "dayjs";
import isInValidRange from "./isValidRange";

const validateAndEmitDate = (
  value: string, 
  format: string, 
  disableDatesBefore?: Date, 
  disableDatesAfter?: Date,
) => {
  const newDateValue = dayjs(value, format);
  if (
    newDateValue.isValid()
    && isInValidRange(
      newDateValue,
      disableDatesBefore ? convertDate(disableDatesBefore) : undefined,
      disableDatesAfter ? convertDate(disableDatesAfter) : undefined,
    )
  ) {
    return newDateValue.format(DATE_FORMAT);
  }
  return undefined;
};

export default validateAndEmitDate;
