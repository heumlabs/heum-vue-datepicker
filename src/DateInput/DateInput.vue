<script lang="ts" setup>
import { ref } from 'vue';

import validateAndEmitDate from './validateAndEmitDate';

const props = defineProps<{
  label: string;
  value: string;
  disableDatesAfter?: Date;
  disableDatesBefore?: Date;
}>();

const emit = defineEmits<{
  (eventName: 'focus', event: Event): void
  (eventName: 'blur', event: Event): void
  (eventName: 'input', value?: string): void
}>();

const inputRef = ref<HTMLInputElement>();
const focused = ref(false);

const handleMouseDownLabel = (e: Event) => {
  if (focused.value) {
    e.preventDefault();
  }
};

const handleFocus = (e: Event) => {
  focused.value = true;
  emit('focus', e);
};

const handleBlur = (e: Event) => {
  focused.value = false;
  emit('blur', e);
};

const handleKeyDownEnter = () => {
  inputRef.value?.blur();
};

const handleClickClear = () => {
  emit('input', undefined);
};

// input handler limit value to numbers and dots
const  handleInput = (e: Event) => {
  const target = e.target as HTMLInputElement;
  target.value = target.value.replace(/([^\d|.])/g, '');
};

// change handler validate date and emit
const handleChange = (e: Event) => {
  const value = (e.target as HTMLInputElement).value;
  const getValidDate = validateAndEmitDate(
    props.disableDatesBefore, 
    props.disableDatesAfter
  )

  const formattedDate = getValidDate(value.slice(0, 8).replace(/\./g, ''), 'YYYYMMDD') 
    || getValidDate(value, 'YYYY.M.D');

  emit('input', formattedDate);
};
</script>

<template>
  <div
    class="date-input"
    :class="{ focused }"
  >
    <label
      @mousedown="handleMouseDownLabel"
    >
      {{ label }}
      <input
        ref="inputRef"
        class="tw-text-base md:tw-text-sm focus:tw-outline-none tw-text-G1 placeholder:tw-text-G3"
        placeholder="YYYY.MM.DD"
        :value="value"
        @input="handleInput"
        @change="handleChange"
        @focus="handleFocus"
        @blur="handleBlur"
        @keydown.enter="handleKeyDownEnter"
      >
    </label>
    <button
      :class="{ hidden: !value }"
      @click="handleClickClear"
    >
      <svg
        width="18"
        height="18"
        viewBox="0 0 18 18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14.25 4.8075L13.1925 3.75L9 7.9425L4.8075 3.75L3.75 4.8075L7.9425 9L3.75
          13.1925L4.8075 14.25L9 10.0575L13.1925 14.25L14.25 13.1925L10.0575 9L14.25 4.8075Z"
        />
      </svg>
    </button>
  </div>
</template>

<style lang="scss" scoped>
@import '@/style.scss';

.date-input {
  box-sizing: border-box;
  display: flex;
  align-items: center;
  position: relative;
  border-radius: 4px;
  border: 1px solid var(--heumlabs-vue-datepicker-disabled-color);
  width: 148px;
  height: 50px;
  padding-left: 16px;
  padding-right: 12px;

  &.focused {
    border-color: var(--heumlabs-vue-datepicker-text-color);
  }

  label {
    min-width: 0;
    flex-grow: 1;
    display: flex;
    flex-direction: column;
    gap: 2px;
    font-size: 10px;
    line-height: 12px;
    color: var(--heumlabs-vue-datepicker-label-color);

    input {
      font-size: 14px;
      line-height: 16px;
      font-weight: bold;
      border: none;
      padding: 0;
      margin: 0;
      outline: none;

      &::placeholder {
        color: var(--heumlabs-vue-datepicker-disabled-color);
      }
    }
  }

  button {
    flex-shrink: 0;
    padding: 0;
    width: 18px;
    height: 18px;

    svg {
      fill: var(--heumlabs-vue-datepicker-text-color);
    }

    &.hidden {
      visibility: hidden;
    }
  }
}
</style>
