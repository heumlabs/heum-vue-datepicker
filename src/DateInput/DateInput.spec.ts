import { flushPromises, mount } from '@vue/test-utils';
import { Dayjs } from 'dayjs';
import DateInput from './DateInput.vue';

interface VM {
  focused: boolean;
}

describe('DateInput.vue', () => {
  it('snapshot tests', () => {
    const wrapper1 = mount(DateInput, {
      props: {
        label: 'DateInput test',
        value: '2022.6.13',
      },
    });

    expect(wrapper1.html()).toMatchSnapshot();

    const wrapper2 = mount(DateInput, {
      props: {
        label: 'DateInput test',
        value: '',
      },
    });

    expect(wrapper2.html()).toMatchSnapshot();
  });

  it('test focused ref value', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'DateInput test',
        value: '',
      },
    });
    expect((wrapper.vm as unknown as VM).focused).toEqual(false);

    const inputElement = wrapper.find('input');
    await inputElement.trigger('focus');
    expect((wrapper.vm as unknown as VM).focused).toEqual(true);

    await inputElement.trigger('blur');
    expect((wrapper.vm as unknown as VM).focused).toEqual(false);
  });

  it('emit focus, blur events', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'DateInput test',
        value: '',
      },
    });
    expect(wrapper.emitted()).not.toHaveProperty('focus');
    expect(wrapper.emitted()).not.toHaveProperty('blur');

    const inputElement = wrapper.find('input');
    await inputElement.trigger('focus');
    expect(wrapper.emitted()).toHaveProperty('focus');
    expect(wrapper.emitted()).not.toHaveProperty('blur');

    await inputElement.trigger('blur');
    expect(wrapper.emitted()).toHaveProperty('focus');
    expect(wrapper.emitted()).toHaveProperty('blur');
  });

  it('blur when keydown enter event triggered', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'DateInput test',
        value: '',
      },
    });

    const inputElement = wrapper.find('input');
    const inputRefBlurMock = vi.spyOn(inputElement.element, 'blur');

    await inputElement.trigger('focus');
    expect(wrapper.emitted()).toHaveProperty('focus');

    await inputElement.trigger('keydown.enter');
    await flushPromises();
    expect(inputRefBlurMock).toHaveBeenCalled();
  });
  it('emit undefined when invalid date input', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'DateInput test',
        value: '',
      },
    });

    const inputElement = wrapper.find('input');
    await inputElement.setValue('2022.234');
    await flushPromises();

    expect((wrapper.emitted().input?.[0] as never)[0]).toEqual('');
  });
  
  it('emit date string when valid date input', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'DateInput test',
        value: '',
      },
    });

    const inputElement = wrapper.find('input');
    await inputElement.setValue('2022.12.1');
    await flushPromises();

    expect(((wrapper.emitted().input?.[0] as never)[0] as Dayjs)).toEqual('2022-12-01');
  });

  it('emit date string when input is valid and date is in valid range', async () => {
    const disableDatesBefore = new Date(2022, 5, 3);
    const disableDatesAfter = new Date(2022, 5, 12);
    const wrapper = mount(DateInput, {
      props: {
        label: 'DateInput test',
        value: '',
        disableDatesAfter,
        disableDatesBefore,
      },
    });

    const inputElement = wrapper.find('input');
    await inputElement.setValue('2022.6.6');
    await flushPromises();

    expect(((wrapper.emitted().input?.[0] as never)[0] as Dayjs)).toEqual('2022-06-06');
  });
  it('emit undefined when input is valid and date is in invalid range 1', async () => {
    const disableDatesBefore = new Date(2022, 5, 3);
    const disableDatesAfter = new Date(2022, 5, 12);
    const wrapper = mount(DateInput, {
      props: {
        label: 'DateInput test',
        value: '',
        disableDatesAfter,
        disableDatesBefore,
      },
    });

    const inputElement = wrapper.find('input');
    await inputElement.setValue('2022.6.1');
    await flushPromises();

    expect((wrapper.emitted().input?.[0] as never)[0]).toEqual('');
  });
  it('emit undefined when input is valid and date is in invalid range 2', async () => {
    const disableDatesBefore = new Date(2022, 5, 3);
    const disableDatesAfter = new Date(2022, 5, 12);
    const wrapper = mount(DateInput, {
      props: {
        label: 'DateInput test',
        value: '',
        disableDatesAfter,
        disableDatesBefore,
      },
    });

    const inputElement = wrapper.find('input');
    await inputElement.setValue('2022.7.1');
    await flushPromises();

    expect((wrapper.emitted().input?.[0] as never)[0]).toEqual('');
  });
  it('유효한 value 값이 있는 경우 clear 버튼이 노출되지 않는다.', () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'DateInput test',
        value: '',
      },
    });

    expect(wrapper.find('button').element.classList.contains('hidden')).toEqual(true);
  });
  it('유효한 value 값이 있는 경우 clear 버튼이 노출된다.', () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'DateInput test',
        value: '2022.6.1',
      },
    });

    expect(wrapper.find('button').exists()).toBeTruthy();
  });
  it('유효한 value 값이 있는 상태에서 clear 버튼 클릭 시 undefined 값을 input 이벤트로 emit 한다.', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'DateInput test',
        value: '2022.6.1',
      },
    });

    const clearButton = wrapper.find('button');
    await clearButton.trigger('click');

    expect((wrapper.emitted().input?.[0] as never)[0]).toEqual('');
  });
});
