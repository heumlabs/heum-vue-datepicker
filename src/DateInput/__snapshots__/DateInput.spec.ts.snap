// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DateInput.vue > snapshot tests 1`] = `
"<div data-v-a3baa652="" class="date-input"><label data-v-a3baa652="">DateInput test <input data-v-a3baa652="" class="tw-text-base md:tw-text-sm focus:tw-outline-none tw-text-G1 placeholder:tw-text-G3" placeholder="YYYY.MM.DD"></label><button data-v-a3baa652="" class=""><svg data-v-a3baa652="" width="18" height="18" viewBox="0 0 18 18" xmlns="http://www.w3.org/2000/svg">
      <path data-v-a3baa652="" d="M14.25 4.8075L13.1925 3.75L9 7.9425L4.8075 3.75L3.75 4.8075L7.9425 9L3.75
          13.1925L4.8075 14.25L9 10.0575L13.1925 14.25L14.25 13.1925L10.0575 9L14.25 4.8075Z"></path>
    </svg></button></div>"
`;

exports[`DateInput.vue > snapshot tests 2`] = `
"<div data-v-a3baa652="" class="date-input"><label data-v-a3baa652="">DateInput test <input data-v-a3baa652="" class="tw-text-base md:tw-text-sm focus:tw-outline-none tw-text-G1 placeholder:tw-text-G3" placeholder="YYYY.MM.DD"></label><button data-v-a3baa652="" class="hidden"><svg data-v-a3baa652="" width="18" height="18" viewBox="0 0 18 18" xmlns="http://www.w3.org/2000/svg">
      <path data-v-a3baa652="" d="M14.25 4.8075L13.1925 3.75L9 7.9425L4.8075 3.75L3.75 4.8075L7.9425 9L3.75
          13.1925L4.8075 14.25L9 10.0575L13.1925 14.25L14.25 13.1925L10.0575 9L14.25 4.8075Z"></path>
    </svg></button></div>"
`;
